neural_lam/weather_dataset.py

```diff
@@ -1,262 +1,373 @@
+                )
+                self.da_forcing_mean = self.ds_forcing_stats.forcing_mean
+                self.da_forcing_std = self.ds_forcing_stats.forcing_std
 
     def __len__(self):
-        return len(self.sample_names)
+        if self.datastore.is_forecast:
+            # for now we simply create a single sample for each analysis time
+            # and then the next ar_steps forecast times
+            if self.datastore.is_ensemble:
+                warnings.warn(
+                    "only using first ensemble member, so dataset size is "
+                    " effectively reduced by the number of ensemble members "
+                    f"({self.da_state.ensemble_member.size})",
+                    UserWarning,
+                )
+            # XXX: we should maybe check that the 2+ar_steps actually fits in
+            # the elapsed_forecast_duration dimension, should that be checked
+            # here?
+            return self.da_state.analysis_time.size
+        else:
+            # sample_len = 2 + ar_steps
+            #             (2 initial states + ar_steps target states)
+            # n_samples = len(self.da_state.time) - sample_len + 1
+            #           = len(self.da_state.time) - 2 - ar_steps + 1
+            #           = len(self.da_state.time) - ar_steps - 1
+            return len(self.da_state.time) - self.ar_steps - 1
+
+    def _sample_time(self, da, idx, n_steps: int, n_timesteps_offset: int = 0):
+        """
+        Produce a time slice of the given dataarray `da` (state or forcing)
+        starting at `idx` and with `n_steps` steps. The `n_timesteps_offset`
+        parameter is used to offset the start of the sample, for example to
+        exclude the first two steps when sampling the forcing data (and to
+        produce the windowing samples of forcing data by increasing the offset
+        for each window).
+
+        Parameters
+        ----------
+        da : xr.DataArray
+            The dataarray to sample from. This is expected to have a `time`
+            dimension if the datastore is providing analysis only data, and a
+            `analysis_time` and `elapsed_forecast_duration` dimensions if the
+            datastore is providing forecast data.
+        idx : int
+            The index of the time step to start the sample from.
+        n_steps : int
+            The number of time steps to include in the sample.
+
+        """
+        # selecting the time slice
+        if self.datastore.is_forecast:
+            # this implies that the data will have both `analysis_time` and
+            # `elapsed_forecast_duration` dimensions for forecasts we for now
+            # simply select a analysis time and then the next ar_steps forecast
+            # times
+            da = da.isel(
+                analysis_time=idx,
+                elapsed_forecast_duration=slice(
+                    n_timesteps_offset, n_steps + n_timesteps_offset
+                ),
+            )
```

Comment:
This now always uses the first `2+ar_steps` time steps of each forecast as the sample. This means that you get 1 training sample per forecast. The other option is to make each lead time in the forecast a separate sample, meaning that you get `forecast_length - ar_steps - 2 ` samples from each forecast (this was the old behavior).

I think this is a perfectly fine change to make, and one can just refine a bit how forecast datasets are used later on. But as this is a change to how this worked previously I think it would be good to mention in a comment here.
```suggestion
            # this implies that the data will have both `analysis_time` and
            # `elapsed_forecast_duration` dimensions for forecasts. We for now
            # simply select a analysis time and the first `n_steps` forecast
            # times (given no offset). Note that this means that we get one sample 
            # per forecast, always starting at forecast time 2.
            da = da.isel(
                analysis_time=idx,
                elapsed_forecast_duration=slice(
                    n_timesteps_offset, n_steps + n_timesteps_offset
                ),
            )
```

Reply:
Looks like github is having issues figuring out this diff, but I only changed the comment :stuck_out_tongue:

Reply:
Ah I see! Yes I wasn't exactly sure how it worked before, but I actually thought this was how it was working but I misunderstood :/ I was going under the assumption that only analysis step (the first step in the forecast) was "a good starting point" for learning how to forecast because the further the forecast has elapsed the further from the true evolution of that atmosphere it will be. This is an interesting point though, I wonder if one trains multiple models on only specific lead times if the final skill of these models will be the same...

Anyway, that is something for later :) I will commit your change to the comment, thank you!